File: Best_Research/#Done/Rust/LearningRust/Units/Physics/src/units/velocity/display.rs
use std::fmt;

use crate::internal::format_unit_value;

use super::quantity::{Velocity, VelocityUnit};

impl Velocity {
    #[must_use]
    pub(crate) const fn display_as(self, unit: VelocityUnit) -> VelocityDisplay {
        VelocityDisplay {
            value: self,
            unit,
            precision: None,
        }
    }

    #[must_use]
    pub(crate) const fn display_as_precision(
        self,
        unit: VelocityUnit,
        precision: usize,
    ) -> VelocityDisplay {
        VelocityDisplay {
            value: self,
            unit,
            precision: Some(precision),
        }
    }

    #[must_use]
    pub const fn display_meters_per_second(self) -> VelocityDisplay {
        self.display_as(VelocityUnit::MetersPerSecond)
    }

    #[must_use]
    pub const fn display_kilometers_per_hour(self) -> VelocityDisplay {
        self.display_as(VelocityUnit::KilometersPerHour)
    }

    #[must_use]
    pub const fn display_meters_per_second_precision(self, precision: usize) -> VelocityDisplay {
        self.display_as_precision(VelocityUnit::MetersPerSecond, precision)
    }

    #[must_use]
    pub const fn display_kilometers_per_hour_precision(self, precision: usize) -> VelocityDisplay {
        self.display_as_precision(VelocityUnit::KilometersPerHour, precision)
    }
}

#[derive(Copy, Clone, Debug)]
pub struct VelocityDisplay {
    value: Velocity,
    unit: VelocityUnit,
    precision: Option<usize>,
}

impl fmt::Display for VelocityDisplay {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        let value = match self.unit {
            VelocityUnit::MetersPerSecond => self.value.to_meters_per_second(),
            VelocityUnit::KilometersPerHour => self.value.to_kilometers_per_hour(),
        };

        format_unit_value(formatter, value, self.unit.symbol(), self.precision)
    }
}

impl fmt::Display for Velocity {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.display_meters_per_second().fmt(formatter)
    }
}

File: Best_Research/#Done/C++/Units-CppXmake/aztro_physics/velocity/display.cpp
module;

#include <format>
#include <iomanip>
#include <limits>
#include <optional>
#include <ostream>

module aztro_physics;

namespace aztro::physics::velocity
{

    const char* symbol(VelocityUnit unit)
    {
        switch (unit)
        {
        case VelocityUnit::MetersPerSecond:
            return "m/s";
        case VelocityUnit::KilometersPerHour:
            return "km/h";
        }

        return "";
    }

    double value_in_unit(Velocity value, VelocityUnit unit)
    {
        switch (unit)
        {
        case VelocityUnit::MetersPerSecond:
            return value.to_meters_per_second();
        case VelocityUnit::KilometersPerHour:
            return value.to_kilometers_per_hour();
        }

        return value.to_meters_per_second();
    }

    QuantityDisplay<Velocity, VelocityUnit> Velocity::display_meters_per_second()
    {
        return display_as(VelocityUnit::MetersPerSecond);
    }

    QuantityDisplay<Velocity, VelocityUnit> Velocity::display_kilometers_per_hour()
    {
        return display_as(VelocityUnit::KilometersPerHour);
    }

    QuantityDisplay<Velocity, VelocityUnit> Velocity::display_meters_per_second_precision(int precision)
    {
        return display_as_precision(VelocityUnit::MetersPerSecond, precision);
    }

    QuantityDisplay<Velocity, VelocityUnit> Velocity::display_kilometers_per_hour_precision(int precision)
    {
        return display_as_precision(VelocityUnit::KilometersPerHour, precision);
    }

    std::ostream& operator<<(std::ostream& stream, Velocity value)
    {
        return stream << value.display_as(VelocityUnit::MetersPerSecond);
    }

} // namespace aztro::physics::velocity
